File: src/main/java/com/wire/xenon/state/JdbiState.java
package com.wire.xenon.state;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.wire.xenon.backend.models.NewBot;
import com.wire.xenon.exceptions.MissingStateException;
import org.jdbi.v3.core.Jdbi;

import java.io.IOException;
import java.util.UUID;

public class JdbiState implements State {
    private final static ObjectMapper mapper = new ObjectMapper();

    private final UUID botId;
    private final StatesDAO statesDAO;

    public JdbiState(UUID botId, Jdbi jdbi) {
        this.botId = botId;
        this.statesDAO = jdbi.onDemand(StatesDAO.class);
    }

    @Override
    public boolean saveState(NewBot newBot) throws IOException {
        String str = mapper.writeValueAsString(newBot);
        return 1 == statesDAO.insert(botId, str);
    }

    @Override
    public NewBot getState() throws IOException {
        String str = statesDAO.get(botId);
        if (str == null)
            throw new MissingStateException(botId);
        return mapper.readValue(str, NewBot.class);
    }

    @Override
    public boolean removeState() {
        return 1 == statesDAO.delete(botId);
    }
}

File: src/main/java/com/wire/kalium/state/JdbiState.kt
package com.wire.kalium.state

import kotlin.Throws
import java.io.IOException
import com.wire.kalium.backend.models.NewBot
import java.util.UUID
import com.fasterxml.jackson.databind.ObjectMapper
import com.wire.kalium.exceptions.MissingStateException
import org.jdbi.v3.core.Jdbi

class JdbiState(private val botId: UUID?, jdbi: Jdbi?) : State {
    private val statesDAO: StatesDAO?
    @Throws(IOException::class)
    override fun saveState(newBot: NewBot?): Boolean {
        val str = mapper.writeValueAsString(newBot)
        return 1 == statesDAO.insert(botId, str)
    }

    @Throws(IOException::class)
    override fun getState(): NewBot? {
        val str = statesDAO.get(botId) ?: throw MissingStateException(botId)
        return mapper.readValue(str, NewBot::class.java)
    }

    override fun removeState(): Boolean {
        return 1 == statesDAO.delete(botId)
    }

    companion object {
        private val mapper: ObjectMapper? = ObjectMapper()
    }

    init {
        statesDAO = jdbi.onDemand(StatesDAO::class.java)
    }
}
